### PR TITLE
SCAL (Dynamic sampling) on docs for simulation

### PR DESCRIPTION
**Pull request by HTAustin**
@nims11  Could you fix the refreshing issue to run scal on docs?

### CODE REVIEW

File: CALEngine/src/bmi_doc_scal.cc [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+        stratums.push_back(vector<int>());
+        for(int doc_id: batch)
+            stratums.back().push_back(doc_id);
+
+        int n = ceil(B*N/(float)T);
+        cerr<<"Sampling "<<n<<" documents"<<endl;
+        vector<int> selector(batch.size());
+        for(int i = 0; i < selector.size(); i++)
+            selector[i] = (i < n?1:0);
+        shuffle(batch.begin(), batch.end(), rand_generator);
```
Comment (sarwar187) [inline CALEngine/src/bmi_doc_scal.cc]: Why are you shuffling here? You already did PRF using selector vector, right?

File: CALEngine/src/bmi_doc_scal.cc [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+        for(int i = 0; i < batch.size(); i++){
+            if(selector[i]) judgment_queue.push_back(batch[i]);
```
Comment (sarwar187) [inline CALEngine/src/bmi_doc_scal.cc]: Here, you are actually doing the PRF, right?

File: CALEngine/src/bmi_doc_scal.cc [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+            cerr<<"Doubling T to "<<T<<endl;
+        }
+        cerr<<"Batch Size = "<<B<<endl;
+        judgments_per_iteration = B;
+        vector<int> batch = perform_training_iteration();
+        stratums.push_back(vector<int>());
+        for(int doc_id: batch)
+            stratums.back().push_back(doc_id);
+
+        int n = ceil(B*N/(float)T);
```
Comment (sarwar187) [inline CALEngine/src/bmi_doc_scal.cc]: I am a little confused here. Is "n" the number of documents that you select for PRF?

File: CALEngine/src/bmi_doc_scal.cc (hunk 2 above)
Comment (HTAustin) [reply CALEngine/src/bmi_doc_scal.cc]: It adds all the judged docs into the training set. For PRF, we need to add the pseudo relevant ones which will be different.

File: CALEngine/src/bmi_doc_scal.cc (hunk 1 above)
Comment (nims11) [reply CALEngine/src/bmi_doc_scal.cc]: Good catch, the selector vector is not doing anything. I don't remember what I was intending to do with it, but it is just useless code now.

File: CALEngine/src/bmi_doc_scal.cc (hunk 3 above)
Comment (nims11) [reply CALEngine/src/bmi_doc_scal.cc]: @sarwar187 we sample `n` documents from the batch to be presented for judgment.

**Pull request merged**